nntrainer/dataset/batch_queue.cpp

```diff
@@ -0,0 +1,68 @@
+// SPDX-License-Identifier: Apache-2.0
+/**
+ * Copyright (C) 2021 Jihoon Lee <jhoon.it.lee@samsung.com>
+ *
+ * @file   batch_queue.cpp
+ * @date   13 July 2021
+ * @brief  This file contains thread safe queue
+ * @see    https://github.com/nnstreamer/nntrainer
+ * @author Jihoon Lee <jhoon.it.lee@samsung.com>
+ * @bug    No known bugs except for NYI items
+ *
+ */
+#include <batch_queue.h>
+#include <chrono>
+
+#include <nntrainer_error.h>
+
+using namespace std::literals::chrono_literals;
+
+namespace nntrainer {
+BatchQueue::BatchQueue(unsigned int queue_capacity_) :
+  queue_capacity(queue_capacity_) {
+  NNTR_THROW_IF(queue_capacity == 0, std::invalid_argument)
+    << "queue capacity of zero not supported!";
+}
+
+BatchQueue::BatchQueue(const BatchQueue &rhs) :
+  queue_capacity(rhs.queue_capacity) {}
+
+BatchQueue &BatchQueue::operator=(const BatchQueue &rhs) {
+  if (this == &rhs) {
+    return *this;
+  }
+  this->queue_capacity = rhs.queue_capacity;
+  return *this;
+}
+
+void BatchQueue::wait_and_push(T &&data) noexcept {
+  std::unique_lock<std::mutex> lk(q_mutex);
+  q_writer_cv.wait(lk, [this] { return q.size() != queue_capacity; });
+  q.push(std::make_unique<T>(data));
+  q_reader_cv.notify_one();
+}
+
+std::unique_ptr<BatchQueue::T> BatchQueue::wait_and_pop() noexcept {
+  std::unique_lock<std::mutex> lk(q_mutex);
+  q_reader_cv.wait(lk, [this] { return !q.empty(); });
+
+  /// @note this invalidates q.front(), but it is okay because it is locked and
+  /// popped right away
+  auto ptr = std::move(q.front());
+  q.pop();
+  q_writer_cv.notify_one();
```

Comment:
It might be good to update this to notify_all() for later support for multiple writers? or add a note/todo.

Reply:
for the mutiple writer, it will be notify_one, because we don't want to wake up every writer when buffer has a room :p

Reply:
I recommend notify_all, as with multiple writers, notify_one might not do anything if all are awake, and lead to some workers sleeping.

Reply:
if all are awake, i think it's intended to be noop? I feel like I am missing something, I'll ask offline :)

Reply:
We discussed offline to add a todo here.

Additional requirements

1. maximize thread utilization
2. resource allocation should be done in parallel
3. minimize overhead